### ISSUE TITLE
Add semantics for CRS

### ISSUE THREAD

**Issue by lvdbrink** [opengeospatial/ogc-geosparql#12]
**Official change request**: [OGC CR 598](http://ogc.standardstracker.org/show_request.cgi?id=598)

Descriptions of Coordinate Reference Systems (CRS) are important in correct interpretation of spatial data, and in working with that kind of data. However, CRS descriptions are not readily available as Linked Data. Therefore, it would be good to have GeoSPARQL use a standard ontology for describing CRS, either as an extension or module of GeoSPARQL itself, or an endorsed external ontology.

**Comment by jabhay** [comment]
MoSCoW poll created

[![](https://api.gh-polls.com/poll/01EJXB6R77B736TK7WGRMZC7AZ/Must)](https://api.gh-polls.com/poll/01EJXB6R77B736TK7WGRMZC7AZ/Must/vote)
[![](https://api.gh-polls.com/poll/01EJXB6R77B736TK7WGRMZC7AZ/Should)](https://api.gh-polls.com/poll/01EJXB6R77B736TK7WGRMZC7AZ/Should/vote)
[![](https://api.gh-polls.com/poll/01EJXB6R77B736TK7WGRMZC7AZ/Could)](https://api.gh-polls.com/poll/01EJXB6R77B736TK7WGRMZC7AZ/Could/vote)
[![](https://api.gh-polls.com/poll/01EJXB6R77B736TK7WGRMZC7AZ/Won%27t)](https://api.gh-polls.com/poll/01EJXB6R77B736TK7WGRMZC7AZ/Won%27t/vote)

**Comment by paulc-dstl** [comment]
EPSG maintain a registry of CRS and CRS components, transformations, etc. See for example the [definition of British National Grid](https://epsg.io/27700) including the WKT specification. The EPSG codes are widely used so could form the basis of Linked Data PURIs.

**Comment by situx** [comment]
Today I have extracted all available proj4 definitions for another purpose. I figured why not converting them to RDF using the same process and try to figure out what would be necessary to create a GeoSPARQL extension ontology as proposed here.
The result is proj4rdf: https://github.com/situx/proj4rdf
It uses pyproj to extract features of the CRS and saves them in three literal types: WKT, Proj4, ProJJSON. It is a start and maybe we can expand upon it.

**Comment by dr-shorthair** [comment]
Is this encoding related to the ISO 19111 model? The CRS types appear to be the same as there.

**Comment by situx** [comment]
I am not really familiar with the ISO 19111 model. 
What I know is:
EPSG Codes are unique identifiers for CRS and that proj4, projjson and WKT are formats that may express the parameters of a CRS.

**Comment by dr-shorthair** [comment]
ISO 19111:2019 is available from OGC here: https://docs.opengeospatial.org/as/18-005r4/18-005r4.html

**Comment by situx** [comment]
After reading the standard, yes these are the terms that are used in the software library as well and which - in any case - should be in an ontology model following this standard.

**Comment by lvdbrink** [comment]
This may also be of interest here: https://github.com/SEMICeu/epsg-to-rdf

From the README:
> This XSLT is a proof of concept for the RDF representation of the OGC EPSG register of coordinate reference systems, extending the RDF mappings for reference systems defined in GeoDCAT-AP

**Comment by FransKnibbe** [comment]
> Hi!
> 
> Here are some additional notes regarding Coordinate Systems (CS') from use cases from the construction industry.
> 
> During the geometry modeling phase for construction works, a custom CS is often used and referenced to an agreed upon CRS. Either approach should differentiate between the applied Coordinate System (e.g. Cartesian CS/ Ellipsoidal CS, Custom CS/CRS, etc.), and propose relations between those coordinate systems, i.e. by defining transformations (translation, rotation, scaling). It might also be interesting to be able to define the units that are used in a coordinate system to allow meaningful analyses of distances, areas, and volumes, as required for quantity surveys.

Yes. An ontology or ontology module for coordinate reference systems should support matryoshka-style superpositions of CRS's.    Assuming both Cartesian and spherical CRS's will be supported, transitions between the too probably will be extra interesting.

**Comment by FransKnibbe** [comment]
Something to look at is the CRS ontology created by IGN France, see http://data.ign.fr/def/ignf/20150505.en.htm.

**Comment by FransKnibbe** [comment]
Very relevant too is the CRS ontology that is being developed by @situx: https://github.com/situx/proj4rdf

**Comment by situx** [comment]
> This may also be of interest here: https://github.com/SEMICeu/epsg-to-rdf
> 
> From the README:
> 
> > This XSLT is a proof of concept for the RDF representation of the OGC EPSG register of coordinate reference systems, extending the RDF mappings for reference systems defined in GeoDCAT-AP

I had a look at this one.
What this XSLT provides that is not available in other libraries that I have seen is a change history of the CRS definition by OGC. If it is important for an ontology to incorporate this information then a conversion to RDF as proposed here would be relevant.

> Something to look at is the CRS ontology created by IGN France, see http://data.ign.fr/def/ignf/20150505.en.htm.

I had a look at this ontology as well. It does not define as many types of CRS as there are included in the ISO 19111 standard which was previously mentioned. Also in some other thread here the question arose whether we should also define non coordinate-based reference systems like what3words. When talking about this, the ontology might as well include geocoding approaches like OpenLocationCode or maybe even (unique) addresses as some form of spatial reference.
However, some things seem to be defined differently in the IGN France ontology compared to my extraction approach and I will form an opinion about this

**Comment by FransKnibbe** [comment]
@situx: What is the main reason for starting proj4rdf? 
And another question: Do you see proj4rdf as a possible ontology for spatial reference systems in general? By that I mean that the ontology could be used to underpin SRSs for e.g. other planets, the solar system, or things like a building, a book page or a computer screen. That way, SRS specifications in different branches of spatial IT could be harmonized. 
Regardless, do you think it's sensible to think up some use cases for an SRS ontology? Those could be used to check all initiatives for capabilities, and see where there is room for improvement.

**Comment by situx** [comment]
@FransKnibbe It started as a project to convert the EPSG database to RDF, which I think is achieved now. Also, it was a way for me to get more familiar with the topic of how to define a CRS.
To answer your questions:
- Yes I would find it very useful to harmonize SRS specifications in IT and for that very reason, I recently added definitions for celestial coordinate systems and also definitions for grid coordinate systems. If you are interested in collaborating in this, we could work on that together
- Regarding use cases: Absolutely, that would be the next step and to be honest, I do not have an overview about those currently. I have just reviewed and integrated definitions and would need to research more to cover use cases.

For me, the question would be where we want to go with this in the working group. Our goal is to broaden the scope of spatial data, so depending on this we could also define the scope of the CRS ontology to include more or fewer contents.

**Comment by FransKnibbe** [comment]
@situx, I think improvement of how SRSs are used is very important. In my experience, when things go wrong with geographic data, on the web or elsewhere, often SRSs are involved. There is also an increasing need to georeference things, and an increasing need for spatial accuracy.
From the perspective of this WG, even if only geographic data are considered, I think the best outcome would be to have one authoritative ontology for SRSs. That way there would be fixed semantics for SRS descriptions, and software would always know how to find the necessary parameters, and how to use them.

Yes, I would like to collaborate on proj4rdf. Maybe the wiki could be used to collect comparisons with similar initiatives and references to existing standards? Also, we could try to make a list of requirements for an SRS ontology (based on use cases).  That kind of information, in turn, could be used to exchange ideas with people behind other SRS initiatives (in order not to reinvent any wheels).

**Comment by situx** [comment]
@FransKnibbe Agreed. Yes, sure we can use the wiki of proj4rdf for that.
I am also organizing my findings in a draft paper currently for eventual publication and to keep track of things.
Important points could be:
- Points or databases of reference in which we can find definitions of reference systems
- Use cases as you state especially but not exclusively from the geospatial community
- And from that, we might as you state infer a list of requirements and add them to the ontology model if we have suggestions on how to meet those requirements.
